rpc/internal/sps.service.go

```diff
@@ -0,0 +1,74 @@
+package internal
+
+import (
+	"time"
+
+	"github.com/evergreen-ci/poplar"
+)
+
+type TaskInformation struct {
+	Project   string `json:"project"`
+	Version   string `json:"version"`
+	Variant   string `json:"variant"`
+	Order     int    `json:"order"`
+	TaskName  string `json:"task_name"`
+	TaskId    string `json:"task_id"`
+	Execution int    `json:"execution"`
+	Mainline  bool   `json:"mainline"`
+}
+
+type TestInformation struct {
+	TestName  string           `json:"test_name"`
+	Arguments map[string]int32 `json:"arguments"`
+}
+
+type TestMetric struct {
+	Name  string      `json:"name"`
+	Value interface{} `json:"value"`
+}
+
+type PerfResult struct {
+	Info        TestInformation `json:"info"`
+	CreatedAt   time.Time       `json:"created_at"`
+	CompletedAt time.Time       `json:"completed_at"`
+	Metrics     []TestMetric    `json:"metrics"`
+}
+
+type SubmitPerfResultRequest struct {
+	Id      TaskInformation `json:"id"`
+	Results []PerfResult    `json:"results"`
+}
+
+func GatherPerfResults(report *poplar.Report) []PerfResult {
+	tests := gatherTests(report.Tests)
+	perfResults := []PerfResult{}
+	for _, test := range tests {
+		testInformation := TestInformation{
+			TestName:  test.Info.TestName,
+			Arguments: test.Info.Arguments,
+		}
+		perfResult := PerfResult{
+			Info:        testInformation,
+			CreatedAt:   test.CreatedAt,
+			CompletedAt: test.CompletedAt,
+			Metrics:     []TestMetric{},
+		}
+		for _, metric := range test.Metrics {
+			perfResult.Metrics = append(perfResult.Metrics, TestMetric{
+				Name:  metric.Name,
+				Value: metric.Value,
+			})
+		}
+		perfResults = append(perfResults, perfResult)
+	}
+	return perfResults
+}
+
+func gatherTests(tests []poplar.Test) []poplar.Test {
+	gatheredTests := []poplar.Test{}
+	for _, test := range tests {
+		gatheredTests = append(gatheredTests, test)
+		gatheredTests = append(gatheredTests, gatherTests(test.SubTests)...)
+	}
+	return gatheredTests
+}
```

Comment:
Starting a random thread to discuss the Evergreen changes:
> The Evergreen codebase will be updated to pull in the new version of Poplar and deal with all of the new config settings. These settings can either be set in a commit or we can pipe them through the actual database configuration system that Evergreen uses. My understanding though is that we can't force restart all hosts and so can't guarantee the time that the hosts will take into account the new config. Is that correct?

I prefer putting settings on the admin page rather than in a commit so they're easier to tweak, but I'm not sure what configuration things you need. If you'll need to tweak them regularly, it might be better for them to not live in our codebase.

I don't have a lot of context on what's happening here, but since @julianedwards has done a couple of migrations he may have advice on how to get all the hosts over to what you want? If we increment the agent version when you commit these changes then they should all roll over to the new version eventually, assuming that's all you need.

Reply:
If SPS is the replacement for Cedar as a centralized "results system" for all performance tests run in Evergreen, then it should be configured at the admin level via the UI. The app servers would pipe the SPS config to task hosts in the same way we pipe the Cedar config. It's a simple model to follow and, as Annie pointed out, we bump the the agent version so there will be a rollover. This is equivalent to a graceful restart: once the running task on each host completes, the app send a signal for the host to restart with the new agent binary so that all "new" perf tasks will be guaranteed to send their results to SPS.

Reply:
>Yeah. It'd be nice to have the hosts changed over at the same time, but if they change over relatively quickly that would be fine. We can handle downstream the small period of time where the hosts are doing different things.

What is your concern about rolling hosts over? We can't kill tasks mid-run (that would be disastrous for users), rolling over ensures that once the currently running task completes, the host is immediately updated. Your changes will be essentially be immediate from the perspective of running performance tasks: once the deploy goes out, all performance tasks started thereafter will start sending their reports to SPS. Furthermore, once the admin SPS config exists in Evergreen, updating it on the admin page will not require a restart of hosts, the new config values will be effective immediately.

Reply:
We'll need to roll over SPS to begin writing to the right place. We'll be testing SPS out by writing to a dummy collection first, but eventually it will need to write to the right place. Ideally at that time, Cedar would no longer be writing anything. Since some of the hosts may already be running, the running tasks will be writing to Cedar and the new tasks will be writing to SPS. 

It's not really a _concern_ more so a potential annoyance if something goes wrong and we need to start debugging.

Reply:
Unless you put a company-wide freeze on all performance tasks I don't think you can avoid this. With that being said, I don't think this will be an annoyance for debugging because you will know which results are going to Cedar and which are going to SPS.

On a related note, are you guys planning on migrating the existing Cedar results to SPS?

Reply:
Yup. The Cedar database and codebase will no longer have anything to do with performance.